File: web-fix/openGoogleMapsDirections.js
/**
 * Open Google Maps directions to a listing location.
 * Prefers the displayed Google/full address; falls back to coordinates with swap correction.
 */
export const normalizeLatLng = (latitude, longitude) => {
  let lat = Number(latitude);
  let lng = Number(longitude);
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) {
    return null;
  }
  if (Math.abs(lat) > 90 || Math.abs(lng) > 180) {
    [lat, lng] = [lng, lat];
  }
  // Common data entry mistake: lat/lng reversed (e.g. India ~26°N, ~71°E stored as 71, 26)
  if (lat >= 68 && lat <= 97 && lng >= 8 && lng <= 37) {
    [lat, lng] = [lng, lat];
  }
  if (lat === 0 && lng === 0) {
    return null;
  }
  return { lat, lng };
};

export const openGoogleMapsDirections = ({
  latitude,
  longitude,
  address = "",
  areaListing = null,
} = {}) => {
  const listing = areaListing || {};
  const addressText = String(
    listing.full_address || address || "",
  ).trim();
  const coords = normalizeLatLng(
    listing.latitude ?? latitude,
    listing.longitude ?? longitude,
  );

  let url = "";
  if (addressText) {
    url = `https://www.google.com/maps/dir/?api=1&destination=${encodeURIComponent(addressText)}`;
  } else if (coords) {
    url = `https://www.google.com/maps/dir/?api=1&destination=${coords.lat},${coords.lng}`;
  } else {
    return false;
  }

  window.open(url, "_blank", "noopener,noreferrer");
  return true;
};

export const openDirectionsForListing = (details) => {
  if (!details) {
    return false;
  }
  return openGoogleMapsDirections({
    latitude: details.latitude,
    longitude: details.longitude,
    address: details.address,
    areaListing: details.area_listing,
  });
};

File: ebroker-sukoon-app/lib/utils/open_google_maps_directions.dart
import 'dart:convert';

import 'package:url_launcher/url_launcher.dart';

class NormalizedLatLng {
  const NormalizedLatLng({required this.lat, required this.lng});

  final double lat;
  final double lng;
}

NormalizedLatLng? listingMapCoordinates({
  required Map<String, dynamic> details,
}) {
  final areaListing = details['area_listing'];
  if (areaListing is Map) {
    final map = Map<String, dynamic>.from(areaListing);
    final lat = double.tryParse('${map['latitude'] ?? map['lat'] ?? ''}');
    final lng = double.tryParse('${map['longitude'] ?? map['lng'] ?? ''}');
    if (lat != null && lng != null && lat != 0 && lng != 0) {
      return NormalizedLatLng(lat: lat, lng: lng);
    }
  }

  final lat = double.tryParse('${details['latitude'] ?? ''}');
  final lng = double.tryParse('${details['longitude'] ?? ''}');
  if (lat != null && lng != null && lat != 0 && lng != 0) {
    return NormalizedLatLng(lat: lat, lng: lng);
  }
  return null;
}

Future<bool> openDirectionsForListing(Map<String, dynamic> details) async {
  final coords = listingMapCoordinates(details: details);
  if (coords == null) return false;

  final uri = Uri.parse(
    'https://www.google.com/maps/dir/?api=1&destination=${coords.lat},${coords.lng}',
  );
  return launchUrl(uri, mode: LaunchMode.externalApplication);
}

String encodeLocationComponents(List<Map<String, dynamic>> components) {
  return jsonEncode(components);
}
